Environment/FilesystemLibrary.cpp

```diff
@@ -1,48 +1,341 @@
+#include <shlwapi.h>
+#include <Scheduler.hpp>
+
+
+std::vector<std::string> disallowed_extensions = {
+    ".exe", ".bat",  ".com", ".cmd",  ".inf", ".ipa",
+".apk", ".apkm", ".osx", ".pif",  ".run", ".wsh",
+".bin", ".app",  ".vb",  ".vbs",  ".scr", ".fap",
+".cpl", ".inf1", ".ins", ".inx",  ".isu", ".job",
+".lnk", ".msi",  ".ps1", ".reg",  ".vbe", ".js",
+".x86", ".pif",  ".xlm", ".scpt", ".out", ".ba_",
+".jar", ".ahk",  ".xbe", ".0xe",  ".u3p", ".bms",
+".jse", ".cpl",  ".ex",  ".osx",  ".rar", ".zip",
+".7z",  ".py",   ".cpp", ".cs",   ".prx", ".tar",
+    ".",  ".wim", ".htm",  ".html", ".css",
+".appimage", ".applescript", ".x86_64", ".x64_64",
+".autorun", ".tmp", ".sys", ".dat", ".ini", ".pol",
+".vbscript", ".gadget", ".workflow", ".script",
+".action", ".command", ".arscript", ".psc1"
+};
+
+const auto pUtilities{Module::Utilities::get_singleton()};
 
 int listfiles(lua_State *L) {
-    const char *directory = luaL_checkstring(L, 1);
+    luaL_checkstring(L, 1);
+
+    pUtilities->create_workspace();
 
-    if (strstr(directory, "..") != nullptr)
+    std::string path = lua_tostring(L, 1);
+
+    if (path.find("..") != std::string::npos)
         luaG_runerrorL(L, "Using \'..\' is not permitted for safety reasons.");
+
+    if(!path.empty()) {
+        path = (pUtilities->location + "\\workspace\\" + path);
+    } else {
+        path = (pUtilities->location + "\\workspace");
+    }
+
+    if (!std::filesystem::is_directory(path.c_str())) {
+        luaG_runerrorL(L, "directory does not exist");
+        return 0;
+    }
+
+    lua_newtable(L);
+
+    int idx = 0;
+    for (const auto& entry : std::filesystem::directory_iterator(path))
+    {
+        auto file_path = entry.path().string();
+        file_path = pUtilities->replace(file_path, (pUtilities->location + "\\workspace\\"), "");
+
+        lua_pushinteger(L, ++idx);
+        lua_pushstring(L, file_path.c_str());
+        lua_settable(L, -3);
+    }
+
     return 1;
 }
 
-void FilesystemLibrary::register_environment(lua_State *L) {
-    // TODO: Implement pipes/websockets with a custom UI for setting the Workspace correctly.
+int writefile(lua_State *L) {
+    luaL_checktype(L, 1, LUA_TSTRING);
+    luaL_checktype(L, 2, LUA_TSTRING);
+    std::string path = lua_tostring(L, 1);
+    std::string content = lua_tostring(L, 2);
 
-    static const luaL_Reg libreg[] = {
-            {"listfiles", listfiles},
-            {nullptr, nullptr},
-    };
+    pUtilities->create_workspace();
 
-    lua_newtable(L);
-    luaL_register(L, nullptr, libreg);
-    lua_setreadonly(L, -1, true);
-    lua_pop(L, 1);
+    std::string extention = PathFindExtensionA(path.c_str());
+    if (extention.empty()) {
+        path += ".txt";
+    } else {
+        for (std::string& test : disallowed_extensions) {
+            if (pUtilities->equals_ignore_case(PathFindExtensionA(path.c_str()), test)) {
+                luaG_runerrorL(L, "attempt to escape directory");
```

Comment:
Change error to "Illegal extension" or a variation. This is not a directory escape.